text/0047-random-assignment-of-availability-chunks.md

```diff
@@ -0,0 +1,283 @@
+CandidateReceipt, the implementation would be too complicated. More details in [Appendix A](#appendix-a).
+- Considering future protocol changes that aim to generalise the work polkadot is doing (like CoreJam), `ParaId`s may be
+removed from the protocol, in favour of more generic primitives. In that case, `ParaId`s in the availability recovery
+process should be replaced with a similar identifier. It's important to note that the implementation is greatly simplified
+if this identifier is part of the `CandidateReceipt` or the future analogous data structure.
+- It's a breaking change that requires most validators to be upgrade their node version.
+
+## Testing, Security, and Privacy
+
+Extensive testing will be conducted - both automated and manual.
+This proposal doesn't affect security or privacy.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+This is a necessary data availability optimisation, as reed-solomon erasure coding has proven to be a top consumer of
+CPU time in polkadot as we scale up the parachain block size and number of availability cores.
+
+With this optimisation, preliminary performance results show that CPU time used for reed-solomon coding can be halved
+and total POV recovery time decrease by 80% for large POVs. See more
+[here](https://github.com/paritytech/polkadot-sdk/issues/598#issuecomment-1792007099).
+
+### Ergonomics
+
+Not applicable.
+
+### Compatibility
+
+This is a breaking change. See [upgrade path](#upgrade-path) section above.
+All validators need to have upgraded their node versions before the feature will be enabled via a runtime upgrade and
+governance call.
+
+## Prior Art and References
+
+See comments on the [tracking issue](https://github.com/paritytech/polkadot-sdk/issues/598) and the
+[in-progress PR](https://github.com/paritytech/polkadot-sdk/pull/1644)
+
+## Unresolved Questions
+
+- Is it the best option to embed the mapping function in the runtime?
+- Is there a better upgrade path that would preserve backwards compatibility?
+- Is usage of `ParaId` the best choice for spreading out the network load during systematic chunk recovery within the
+same block?
+
+## Future Directions and Related Material
+
+This enables future optimisations for the performance of availability recovery, such as retrieving batched systematic
+chunks from backers/approval-checkers.
+
+## Appendix A
+
+This appendix explores alternatives to using the `ParaId` as the factor by which availability chunk indices are
+distributed to validators within the same relay chain block, and why they weren't chosen.
+
+### Core index
+
+Here, `core_index` refers to the index of the core that a candidate was occupying while it was pending availability
+(from backing to inclusion).
+
+Availability-recovery can currently be triggered by the following phases in the polkadot protocol:
+1. During the approval voting process.
```

Comment:
As long as recovery is possible in disputes, it should be fine if it can not always use systemic recovery. What is missing for me, is to understand why this needs to be an incompatible change. Shouldn't recovery be possible always, even if you don't know what the systemic chunks are?

Reply:
Currently, when you request a chunk, you need to know which chunk to request. If you specify a wrong chunk, you'll get an empty response. So everyone should be onboard how chunks are shuffled in order for reconstruction to work, not just from systematic chunks. Unless we rely on backers, which we can't in case of a dispute.

We could add an additional request type to request any chunk you have. That could probably be a reasonable fallback in this case.

Reply:
> We could add an additional request type to request any chunk you have. That could probably be a reasonable fallback in this case.

I don't think this achieves the purpose of availability-recovery working with an arbitrary mix of upgraded & non-upgraded validators. Adding a new request type (or even changing the meaning of the existing one) would also mean a node upgrade (because responding to the new type of request is only useful for the nodes that haven't yet upgraded to use the new mapping). So validators might as well just upgrade to the version that uses the new shuffle. I'm not sure this has any benefit

Reply:
I meant as a fallback in case we want to use `core_id`s which might be not readily available in some cases of disputes.

Reply:
> I meant as a fallback in case we want to use core_ids which might be not readily available in some cases of disputes.

I see. In this case, it would be a reasonable fallback. But this would mean giving up systematic recovery in those cases. I'm wondering if it's worth doing all of this just to replace the para-id thingy with the core_id

Reply:
As discussed in chat. If we keep requesting via `ValidatorIndex`, then in recovery you don't need to know the mapping, if you are fine with non systemic recovery.

Reply:
You need to know the chunk-validator mapping.  we may say the chunk index in the chunk itself, but you'd then think you could still get the same chunk for the guy who just gave you one.  In extreme cases, I bet this breaks reconstruction.

Reply:
Indeed, we discussed this. The only issue is that you could not know whether you got the chunk you actually asked for. But that is only relevant if we wanted to change reputations. With that aside, we can just check whether we already fetched that chunk - if yes, treat it as invalid/garbage and move on to the next validator.

Reply:
It's worse than reputations.  I think it changes whether the 2/3rds honest assumption works correctly.

Reply:
> It's worse than reputations. I think it changes whether the 2/3rds honest assumption works correctly.

Hm, how?

we currently don't modify any reputations within the availability-recovery subsystem. also, what @eskimor is describing above can happen with the current production code. the node does not check that it received the chunk it requested. It only checks that it's a valid chunk and that the proof is valid

Reply:
The first 1/3rd of validators refuse to talk to you.  You obtain some chunks from most of the second 1/3rd, but they give you the chunks of the third 1/3rd, so you keep asking the second 1/3rd for the last chunks, which they do not have.  Reconstruction fails.

Reply:
> You obtain some chunks from most of the second 1/3rd, but they give you the chunks of the third 1/3rd

Doesn't that mean the second third is also malicious? They don't adhere to the network protocol (which in polkadot's case is still specified by the reference implementation AFAICT).

We assume that (aside from the malicious third), all validators will respond with unique chunks. We don't care that they're not the ones they should/shouldn't hold, as long as they don't overlap. I think this is a reasonable assumption, isn't it? (we are already making this assumption in the current released code).

Leaving systematic recovery aside, the key aspect would be that we wouldn't ask for the chunk by its index. we would ask each validator for their chunk. The responders would map their validator index to the right chunk index. They would then respond with that chunk. Assuming at most a third of them are malicious and they all have the same mapping, we should get at least two thirds correct responses (and we only need a third. And for non-systematic recovery, we don't care if the responses are actually the ones we asked for, as long as we don't get duplicates.

For systematic recovery, we'd assume the validator initiating recovery has the mapping and requests chunks from the validators. Systematic recovery will then check that the chunks received are indeed the ones requested and fall back to regular behaviour if a peer misbehaves.

Reply:
> No. We're discussing if it's okay for validators to have different mappings. I'm saying no.

I was discussing the case where validators all have the same mapping, but a validator participating in a dispute does not have access to the mapping at that point.

Reply:
Ahh fine.  We'll loose the mappings sometime before the data expires?  We should probably avoid doing that.  We can obviously ask everyone and account for the DLs differently, but I'd prefer to be more like bittorrent trackers here, so that parachains can fetch data this way.

Reply:
Yes. All validators will know the correct mapping, if they have seen (and still know) the relay parent. There is a good chance that this is the case even in disputes, but it is not guaranteed and recovery must work even in that case (but can be slower).

Using the core index should be fine, what makes that even more acceptable to me is that we actually want (and need to) include the core index in the candidate receipt at some point. So the moment we have this change, then the block availability will cease to matter.